File: scripts/perf/catalog_bytes.py
"""Compact JSON byte counts for catalogs. Not the same as HTML/RSC transfer."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Mapping

from scripts.perf.profiles import LOGIN_NAMESPACES, OVERVIEW_NAMESPACES

FRONTEND_ROOT = Path(__file__).resolve().parents[2] / "frontend" / "src" / "i18n" / "messages"


def compact_bytes(value: object) -> int:
    return len(json.dumps(value, ensure_ascii=False, separators=(",", ":")).encode("utf-8"))


def load_catalog(locale: str, root: Path | None = None) -> dict[str, Any]:
    base = root or FRONTEND_ROOT
    payload = _read_json(base / f"{locale}.json")
    monitor = _read_json(base / "monitor" / f"{locale}.json")
    settings = payload.get("settings")
    merged_settings = dict(settings) if isinstance(settings, dict) else {}
    merged_settings["monitor"] = monitor
    payload["settings"] = merged_settings
    return payload


def pick_namespaces(catalog: Mapping[str, Any], names: tuple[str, ...]) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for name in names:
        _assign_path(result, catalog, name.split("."), name)
    return result


def _assign_path(
    target: dict[str, Any],
    source: Mapping[str, Any],
    parts: list[str],
    namespace: str,
) -> None:
    if not parts:
        raise KeyError(f"Missing message namespace: {namespace}")
    head, *rest = parts
    if head not in source:
        raise KeyError(f"Missing message namespace: {namespace}")
    if not rest:
        target[head] = source[head]
        return
    value = source[head]
    if not isinstance(value, Mapping):
        raise TypeError(f"Message namespace is not nested: {namespace}")
    child = target.get(head)
    if not isinstance(child, dict):
        child = {}
        target[head] = child
    _assign_path(child, value, rest, namespace)


def catalog_byte_report(root: Path | None = None) -> dict[str, Any]:
    locales = {}
    for locale in ("en-US", "zh-CN"):
        catalog = load_catalog(locale, root)
        namespaces = {
            name: compact_bytes(value)
            for name, value in catalog.items()
            if isinstance(value, (dict, str, list))
        }
        locales[locale] = {
            "full": compact_bytes(catalog),
            "login_subset": compact_bytes(pick_namespaces(catalog, LOGIN_NAMESPACES)),
            "overview_subset": compact_bytes(pick_namespaces(catalog, OVERVIEW_NAMESPACES)),
            "namespaces": namespaces,
        }
    return {
        "note": (
            "Compact UTF-8 JSON of the server catalog. This is not HTML, RSC, "
            "or gzip transfer size; those belong in the browser section."
        ),
        "locales": locales,
        "client_estimate": {
            "login": locales["en-US"]["login_subset"],
            "overview": locales["en-US"]["overview_subset"],
            "login_zh": locales["zh-CN"]["login_subset"],
            "overview_zh": locales["zh-CN"]["overview_subset"],
        },
    }


def _read_json(path: Path) -> dict[str, Any]:
    payload = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(payload, dict):
        raise ValueError(f"{path} must contain a JSON object")
    return payload

File: frontend/scripts/catalog-bytes.mjs
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

/** Keep in sync with frontend/src/i18n/namespaces.ts client payloads. */
export const LOGIN_NAMESPACES = ["feedback", "login"];
export const OVERVIEW_NAMESPACES = [
  "feedback",
  "site",
  "nav",
  "shell",
  "serverDetail",
  "plugins.aiImport",
];

const here = dirname(fileURLToPath(import.meta.url));
const messagesRoot = join(here, "../src/i18n/messages");

export function compactBytes(value) {
  return Buffer.byteLength(JSON.stringify(value), "utf8");
}

function isPlainObject(value) {
  return value != null && typeof value === "object" && !Array.isArray(value);
}

function assignPath(target, source, parts, namespace) {
  const head = parts[0];
  if (head == null || !(head in source)) {
    throw new Error(`Missing message namespace: ${namespace}`);
  }
  if (parts.length === 1) {
    target[head] = source[head];
    return;
  }
  const value = source[head];
  if (!isPlainObject(value)) {
    throw new Error(`Message namespace is not nested: ${namespace}`);
  }
  const next = isPlainObject(target[head]) ? target[head] : {};
  target[head] = next;
  assignPath(next, value, parts.slice(1), namespace);
}

export function pickNamespaces(catalog, names) {
  const result = {};
  for (const name of names) {
    assignPath(result, catalog, name.split("."), name);
  }
  return result;
}

export function loadCatalog(locale) {
  const main = JSON.parse(readFileSync(join(messagesRoot, `${locale}.json`), "utf8"));
  const monitor = JSON.parse(readFileSync(join(messagesRoot, "monitor", `${locale}.json`), "utf8"));
  return { ...main, settings: { ...(main.settings ?? {}), monitor } };
}

export function catalogByteReport() {
  const locales = {};
  for (const locale of ["en-US", "zh-CN"]) {
    const catalog = loadCatalog(locale);
    locales[locale] = {
      full: compactBytes(catalog),
      login_subset: compactBytes(pickNamespaces(catalog, LOGIN_NAMESPACES)),
      overview_subset: compactBytes(pickNamespaces(catalog, OVERVIEW_NAMESPACES)),
    };
  }
  return {
    note: "Compact UTF-8 JSON of the server catalog, not HTML/RSC/gzip transfer.",
    locales,
    client_estimate: {
      login: locales["en-US"].login_subset,
      overview: locales["en-US"].overview_subset,
      login_zh: locales["zh-CN"].login_subset,
      overview_zh: locales["zh-CN"].overview_subset,
    },
  };
}

if (import.meta.url === `file://${process.argv[1]}`) {
  process.stdout.write(`${JSON.stringify(catalogByteReport(), null, 2)}\n`);
}
